src/Field.php

```diff
@@ -489,7 +489,12 @@ public function setNull(): self
      */
     public function compare($value): bool
     {
-        return $this->get() == $value;
+        if ($this->owner->persistence === null) {
+            return (string) $this->normalize($this->get()) === (string) $this->normalize($value);
+        }
+
+        return (string) $this->owner->persistence->typecastSaveRow($this->owner, [$this->short_name => $this->get()])[$this->short_name]
```

Comment:
this will fail if `get()` does not use `->id` (see https://github.com/atk4/data/pull/707/files#r483735690 )